test/e2e/e2e_suite_test.go

```diff
@@ -73,7 +73,7 @@ var _ = SynchronizedBeforeSuite(func() []byte {
 	e2eConfig = LoadE2EConfig(configPath)
 
 	if useExistingCluster {
-		kubeconfigPath = os.Getenv("KUBECONFIG")
+		kubeconfigPath = os.Getenv("KUBECONFIG_E2E")
```

Comment:
I don't think we should change this. It will be very surprising that we do not use the standard environment variable.

Reply:
Hmm. That's true, but I don't know if it's supported in CAPI framework. I can investigate.
Anw, could you elaborate why not using standard env var is a bad thing here?

Reply:
It has nothing to do with CAPI. It is standard in Kubernetes itself, e.g. `kubectl` use it and most other CLIs that deal with the Kubernetes API, like `clusterctl`, `cmctl` and `argocd`.
Since this has become a standard that so many use, I would as a user expect that the same variable is used also here. Using the standard variable means less need for documenting the behavior since it "works as expected". If we use a custom variable it must definitely be documented.

If you need an example for how to implement this with contexts, check CAPO.
- [Here](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/fc7ba6edd13fe05dec9431c643c710a45dbf4f31/test/e2e/data/e2e_conf.yaml#L166) is the variable for what context to use
- [This is how they use it](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/fc7ba6edd13fe05dec9431c643c710a45dbf4f31/test/e2e/shared/cluster.go#L72)

Reply:
Hmm. Sorry I'm still not following :sweat_smile: The example you showed seems to just check if the context is correct, and does nothing if it's not.
In our case, we will have to switch back and forth between the contexts.

My point regarding CAPI is that `framework.NewClusterProxy()` just takes a kubeconfig and doesn't care about the context.
And don't we get non-default kubeconfig files anyway if we create clusters with `bootstrap.CreateKindBootstrapClusterAndLoadImages()`?

Reply:
> In our case, we will have to switch back and forth between the contexts.

Good point. It may be easier with two different kubeconfigs.

> And don't we get non-default kubeconfig files anyway if we create clusters with bootstrap.CreateKindBootstrapClusterAndLoadImages()?

The point is not about where the file is, it is that `KUBECONFIG` is a standard environment variable. I think it is better that we keep it as is. For the new kubeconfig used for upgrade, you can add another variable. However, this *must* be in the e2e config, just like all other variables we add.
I would also suggest that if this "BMO upgrade" variable is not set, we should fallback to `KUBECONFIG` also in this case. It will probably not work in most cases, but it would be the expected "default". In some cases it could even work, e.g. if the user was running only the BMO upgrade test.

Reply:
That makes senses, but a bit tricky to implement. `ci-e2e.sh` needs to know this config to be able to populate the cluster, but we've also discussed that it shouldn't read the config file. I think we need a new solution for this thing :-?

Reply:
Think about `ci-e2e.sh` as "the user". It can create the cluster and configure the tests to use it. It does not need to know where the kubeconfig is, because it creates it and decides where to put it. In pseudo code

```
# ci-e2e.sh
create_bmo_upgrade_cluster()
save_kubeconfig(/tmp/bmo-upgrade-kubeconfig.yaml)

export BMO_UPGRADE_KUBECONFIG=/tmp/bmo-upgrade-kubeconfig.yaml
```

```yaml
# e2e-config.yaml
variables:
  BMO_UPGRADE_KUBECONFIG: "overridden by environment variable"
```

```
# In test code
if config:= GetVariable(BMO_UPGRADE_KUBECONFIG)
  use_kubeconfig(config)
else
  use_kubeconfig(KUBECONFIG if defined, otherwise ~/.kube/config)

```